### PR TITLE
Adopt concept of flows in Microsoft Auth

### LINKED ISSUE #236825: [MSAL] Can't sign in in a browser codespace
Does this issue occur when all extensions are disabled?: I don't know how to start with --disable-extensions in a codespace, but I don't expect that to matter.

Version: 1.96.2
Commit: fabdb6a30b49f79a7aba0f2ad9df9b399473380f
Date: 2024-12-19T10:22:47.216Z
Browser: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36 Edg/133.0.0.0

Sorry, here's another one...

Steps to Reproduce:

1. Open a GitHub Codespace in the browser
2. Set `"microsoft-authentication.implementation": "msal"`
3. Try to sign in using a Microsoft account (Sync Settings, Azure Repos, etc)

Expected: I can sign in

Actual: After signing in on the new browser tab it creates, I get redirected to a localhost page that doesn't connect, and the Codespace continues to say "Signing in to Microsoft...". The backup protocol handler option also doesn't seem to do anything.

localhost-not-available

(I'm not sure if this is a supported MSAL scenario yet? #178740 only discusses vscode.dev so I can't tell)

<details>
<summary>Microsoft Authentication</summary>
<pre>
2024-12-22 19:41:11.070 [info] Acquired MSAL enablement value from 'globalValue'. Value: msal
2024-12-22 19:41:11.070 [debug] [initialize] Initializing PublicClientApplicationManager
2024-12-22 19:41:12.904 [warning] [Sun, 22 Dec 2024 19:41:12 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Warning - NativeBroker implementation was provided but the broker is unavailable.
2024-12-22 19:41:12.904 [debug] [update] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] CachedPublicClientApplication update before: 0
2024-12-22 19:41:12.904 [debug] [Sun, 22 Dec 2024 19:41:12 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Info - getTokenCache called
2024-12-22 19:41:13.375 [debug] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-common@14.16.0 : Info - CacheManager:getIdToken - Returning ID token
2024-12-22 19:41:13.376 [debug] [update] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] CachedPublicClientApplication update after: 1
2024-12-22 19:41:13.376 [debug] [update] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] CachedPublicClientApplication accounts changed. added: 1, deleted: 0
2024-12-22 19:41:13.376 [debug] [update] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] CachedPublicClientApplication update complete
2024-12-22 19:41:13.380 [debug] [initialize] PublicClientApplicationManager initialized
2024-12-22 19:41:13.381 [debug] [_handleAccountChange] added: 1, changed: 0, deleted: 0
2024-12-22 19:41:13.709 [info] [getSessions] [all] starting
2024-12-22 19:41:13.710 [info] [getSessions] [all] returned 1 session(s)
2024-12-22 19:41:13.710 [info] [getSessions] [all] starting
2024-12-22 19:41:13.710 [info] [getSessions] [all] returned 1 session(s)
2024-12-22 19:41:13.710 [info] [getSessions] [email offline_access openid profile] starting
2024-12-22 19:41:13.712 [debug] [getOrCreate] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] PublicClientApplicationManager cache hit
2024-12-22 19:41:13.713 [debug] [acquireTokenSilent] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] [email offline_access openid profile User.Read] [winstonliu@microsoft.com] starting...
2024-12-22 19:41:13.714 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - initializeRequestScopes called
2024-12-22 19:41:13.716 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [48d59f39-6fb3-49a3-b360-2e79a470cd68] : @azure/msal-node@2.16.2 : Verbose - buildOauthClientConfiguration called
2024-12-22 19:41:13.716 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [48d59f39-6fb3-49a3-b360-2e79a470cd68] : @azure/msal-node@2.16.2 : Verbose - createAuthority called
2024-12-22 19:41:13.717 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Attempting to get cloud discovery metadata  from authority configuration
2024-12-22 19:41:13.717 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Did not find cloud discovery metadata in the config... Attempting to get cloud discovery metadata from the hardcoded values.
2024-12-22 19:41:13.717 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Found cloud discovery metadata from hardcoded values.
2024-12-22 19:41:13.718 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Attempting to get endpoint metadata from authority configuration
2024-12-22 19:41:13.718 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Did not find endpoint metadata in the config... Attempting to get endpoint metadata from the hardcoded values.
2024-12-22 19:41:13.719 [debug] [Sun, 22 Dec 2024 19:41:13 GMT] : [48d59f39-6fb3-49a3-b360-2e79a470cd68] : @azure/msal-node@2.16.2 : Info - Building oauth client configuration with the following authority: https://login.microsoftonline.com/organizations/oauth2/v2.0/token.
2024-12-22 19:41:13.723 [trace] [Sun, 22 Dec 2024 19:41:13 GMT] : [48d59f39-6fb3-49a3-b360-2e79a470cd68] : @azure/msal-node@2.16.2 : Verbose - Silent flow client created
2024-12-22 19:41:13.724 [debug] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-common@14.16.0 : Info - CacheManager:getAccessToken - Returning access token
2024-12-22 19:41:13.724 [debug] [Sun, 22 Dec 2024 19:41:13 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-common@14.16.0 : Info - CacheManager:getIdToken - Returning ID token
2024-12-22 19:41:13.724 [debug] [acquireTokenSilent] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] [email offline_access openid profile User.Read] [winstonliu@microsoft.com] got result
2024-12-22 19:41:13.724 [info] [getSessions] [email offline_access openid profile] returned 1 session(s)
2024-12-22 19:41:14.284 [info] [getSessions] [all] starting
2024-12-22 19:41:14.284 [info] [getSessions] [all] returned 1 session(s)
2024-12-22 19:41:36.298 [info] [createSession] [email offline_access openid profile] starting
2024-12-22 19:41:36.298 [debug] [getOrCreate] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] PublicClientApplicationManager cache hit
2024-12-22 19:41:36.314 [debug] [acquireTokenInteractive] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] [email offline_access openid profile User.Read] loopbackClientOverride: false
2024-12-22 19:41:36.517 [debug] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-node@2.16.2 : Info - getAuthCodeUrl called
2024-12-22 19:41:36.517 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-node@2.16.2 : Verbose - initializeRequestScopes called
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-node@2.16.2 : Verbose - buildOauthClientConfiguration called
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-node@2.16.2 : Verbose - createAuthority called
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Attempting to get cloud discovery metadata  from authority configuration
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Did not find cloud discovery metadata in the config... Attempting to get cloud discovery metadata from the hardcoded values.
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Found cloud discovery metadata from hardcoded values.
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Attempting to get endpoint metadata from authority configuration
2024-12-22 19:41:36.518 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] : @azure/msal-node@2.16.2 : Verbose - Did not find endpoint metadata in the config... Attempting to get endpoint metadata from the hardcoded values.
2024-12-22 19:41:36.518 [debug] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-node@2.16.2 : Info - Building oauth client configuration with the following authority: https://login.microsoftonline.com/organizations/oauth2/v2.0/token.
2024-12-22 19:41:36.519 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-node@2.16.2 : Verbose - Auth code client created
2024-12-22 19:41:36.519 [trace] [Sun, 22 Dec 2024 19:41:36 GMT] : [08dfb6d9-80cc-422f-8223-e98075fba9c9] : @azure/msal-common@14.16.0 : Verbose - createAuthCodeUrlQueryString: Prompt is select_account, ignoring account hints
2024-12-22 19:41:49.460 [debug] [acquireTokenInteractive] [aebc6443-996d-45c2-90f0-388ff96faa56] [https://login.microsoftonline.com/organizations] [email offline_access openid profile User.Read] loopbackClientOverride: true
</pre>
</details>

<details>
<summary>Window</summary>
<pre>
2024-12-22 14:40:47.789 [info] Resolving connection token (codespaces)...
2024-12-22 14:40:47.857 [warning] Via 'product.json#extensionEnabledApiProposals' extension 'ms-vscode.vscode-selfhost-test-provider' wants API proposal 'attributableCoverage' but that proposal DOES NOT EXIST. Likely, the proposal has been finalized (check 'vscode.d.ts') or was abandoned.
2024-12-22 14:40:47.857 [warning] Via 'product.json#extensionEnabledApiProposals' extension 'ms-python.python' wants API proposal 'contribIssueReporter' but that proposal DOES NOT EXIST. Likely, the proposal has been finalized (check 'vscode.d.ts') or was abandoned.
2024-12-22 14:40:47.857 [warning] Via 'product.json#extensionEnabledApiProposals' extension 'ms-python.debugpy' wants API proposal 'contribIssueReporter' but that proposal DOES NOT EXIST. Likely, the proposal has been finalized (check 'vscode.d.ts') or was abandoned.
2024-12-22 14:40:53.376 [info] Updating additional builtin extensions cache
2024-12-22 14:41:00.885 [info] Resolved connection token (codespaces) after 13096 ms
2024-12-22 14:41:00.887 [info] Creating a socket (renderer-Management-f190c0cc-251d-47e3-bbac-00273f7615b8)...
2024-12-22 14:41:00.897 [info] Creating a socket (renderer-ExtensionHost-8e1b0fdc-9db7-4fcf-969f-1f32c201bdc8)...
2024-12-22 14:41:01.216 [info] Creating a socket (renderer-Management-f190c0cc-251d-47e3-bbac-00273f7615b8) was successful after 329 ms.
2024-12-22 14:41:01.273 [info] Creating a socket (renderer-ExtensionHost-8e1b0fdc-9db7-4fcf-969f-1f32c201bdc8) was successful after 376 ms.
2024-12-22 14:41:04.105 [error] Extension 'ms-vscode.vscode-selfhost-test-provider' wants API proposal 'attributableCoverage' but that proposal DOES NOT EXIST. Likely, the proposal has been finalized (check 'vscode.d.ts') or was abandoned.
2024-12-22 14:41:04.222 [warning] [twxs.cmake]: Cannot register 'cmake.cmakePath'. This property is already registered.
2024-12-22 14:41:14.610 [info] Settings Sync: Account status changed from uninitialized to unavailable
2024-12-22 14:41:22.799 [info] Extension host (Remote) is unresponsive.
2024-12-22 14:41:24.182 [info] Extension host (Remote) is responsive.
2024-12-22 14:41:31.683 [error] [Extension Host] (node:686) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
</pre>
</details>

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
And only use Loopback flow when not running in Remote Extension Host.

Fixes https://github.com/microsoft/vscode/issues/236825

### CODE REVIEW

**Review by 818Nawaf — APPROVED** [review]

File: extensions/microsoft-authentication/src/node/authProvider.ts [hunk 1]
```diff
@@ -187,86 +186,70 @@ export class MsalAuthProvider implements AuthenticationProvider {
...
 			}
+		};
 
-			// The user closed the modal window
-			if ((e as ClientAuthError).errorCode === ClientAuthErrorCodes.userCanceled) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
-			}
+		const flows = getMsalFlows({
+			extensionHost: typeof navigator === 'undefined'
```
Comment (legomushroom) [inline extensions/microsoft-authentication/src/node/authProvider.ts:209]: 💭 Nested ternary expressions a bit hard to read especially when inside the other expression, would it be better to have a `extensionHost` variable for it declared above?

File: extensions/microsoft-authentication/src/node/authProvider.ts [hunk 2]
```diff
@@ -187,86 +186,70 @@ export class MsalAuthProvider implements AuthenticationProvider {
...
+				? this._context.extension.extensionKind === ExtensionKind.UI ? ExtensionHost.Local : ExtensionHost.Remote
+				: ExtensionHost.WebWorker,
+		});
 
-			// The user wants to try the loopback client or we got an error likely due to spinning up the server
-			const loopbackClient = new UriHandlerLoopbackClient(this._uriHandler, redirectUri, this._logger);
+		let lastError: Error | undefined;
+		for (const flow of flows) {
+			if (flow !== flows[0]) {
```
Comment (legomushroom) [inline extensions/microsoft-authentication/src/node/authProvider.ts:216]: What does this do? Does it check that a correct logic flow was used?

File: extensions/microsoft-authentication/src/node/authProvider.ts [hunk 3]
```diff
@@ -187,86 +186,70 @@ export class MsalAuthProvider implements AuthenticationProvider {
...
+				const session = this.sessionFromAuthenticationResult(result, scopeData.originalScopes);
+				this._telemetryReporter.sendLoginEvent(session.scopes);
+				this._logger.info('[createSession]', `[${scopeData.scopeStr}]`, 'returned session');
+				this._onDidChangeSessionsEmitter.fire({ added: [session], changed: [], removed: [] });
+				return session;
 			} catch (e) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
+				lastError = e;
+				if (e instanceof ServerError || (e as ClientAuthError)?.errorCode === ClientAuthErrorCodes.userCanceled) {
```
Comment (legomushroom) [inline extensions/microsoft-authentication/src/node/authProvider.ts:240]: The `as` type casting is dangerous and imho must be forbidden by the linter ;) Is it possible to do the same without the casing?

File: extensions/microsoft-authentication/src/node/authProvider.ts [hunk 4]
```diff
@@ -187,86 +186,70 @@ export class MsalAuthProvider implements AuthenticationProvider {
...
+					this._telemetryReporter.sendLoginFailedEvent();
+					throw e;
+				}
+				// Continue to next flow
+				if (e instanceof CancellationError) {
```
Comment (legomushroom) [inline extensions/microsoft-authentication/src/node/authProvider.ts:245]: Nit: could be

```suggestion
			userCancelled = (e instanceof CancellationError);
```

Also do we need to re-throw in the cancellation case too?

**Review by legomushroom — APPROVED** [review]
LGTM 🚀

**Comment by kzhui125** [comment]
@legomushroom Can you check this issue? https://github.com/microsoft/vscode/issues/222022

**Comment by kzhui125** [comment]
@legomushroom Can you check this issue? https://github.com/microsoft/vscode/issues/222022

**Pull request merged**